Hw2database

Boroda9992 opened the pull request:
Не работают тесты. Нужна помощь с аннотацией @Autowired. В материалах курса не смог найти конкретное место, где разбирается ее использование. Официальная документация не особо пролила свет на работу с базами данных. 
По суть, код выдает ошибку: 
java.lang.NullPointerException: Cannot invoke "ru.hogwarts.school.repositories.FacultyRepository.save(Object)" because "this.facultyRepository" is null
Я попробовал добавить @Autowired на инжекцию репозитория в имплементации сервиса, однако это результата не дало (даже негативного - мне не подчеркнуло ошибку).

Boroda9992 committed:
Как обычно, я забыл с самого начала создать ветку и просто начал писать код. Надеюсь, что пулл-реквест будет достаточно читаемым.

Boroda9992 committed:
Тесты не работают. Где-то я должен подключить @Autowired, однако, я не понимаю где и как.

Antony8720 commented on src/test/java/ru/hogwarts/school/service/FacultyServiceImplTest.java:
```diff
@@ -0,0 +1,62 @@
+import ru.hogwarts.school.model.Faculty;
+import ru.hogwarts.school.repositories.FacultyRepository;
+
+public class FacultyServiceImplTest {
```
В данном тесте, поскольку у тебя класс сервиса внутри себя вызывает класс репозитория, необходимо использовать mockito для того, чтобы замокать репозиторий в тестах сервиса, как это делать подробно объяснялось в уроке по теме mockito, можешь посмотреть в его теории, как это использовать, еще раз. 
Здесь необходимо использовать аннтотации @Mock и @InjectMocks.

Boroda9992 replied:
сделано

Antony8720 commented on src/test/java/ru/hogwarts/school/service/StudentServiceImplTest.java:
```diff
@@ -0,0 +1,65 @@
+import ru.hogwarts.school.repositories.StudentRepository;
+
+public class StudentServiceImplTest {
+    private StudentRepository studentRepository;
```
Здесь аналогично прошлому комментарию

Boroda9992 replied:
сделано

Boroda9992 committed:
Починил тесты. Так и не понял где нужно использовать аннотацию @Autowired.

Antony8720 commented on src/test/java/ru/hogwarts/school/service/FacultyServiceImplTest.java:
```diff
@@ -0,0 +1,70 @@
+        Faculty facultyBeingAdded = new Faculty(facultyToAdd.getId(), facultyToAdd.getName(), facultyToAdd.getColour());
+
+        //then
+        Assertions.assertEquals(facultyToAdd, facultyBeingAdded);
```
Тесты написаны не совсем верно.

Сейчас данный тест проверяет только то, что ты корректно руками создал два объекта - facultyToAdd и facultyBeingAdded. Здесь не проверятся корректная работа метода сервиса. 

Данный тест должен выглядеть примерно так:

Задается объект для добавления

Поскольку facultyService внутри себя вызывает FacultyRepository, необходимо замокать вызов метода facultyRepoitory.save() с необходимыми входными параметрами, чтобы мокито его подменил
`Mockito.when(facultyRepository.save(объект для добавления).thenReturn(необходимый объект);`
Подробнее можешь посмотреть тут https://habr.com/ru/companies/otus/articles/780090/

Далее нужно сравнить ожидаемое значение и значение, полученное в результате вызова метода addFaculty (Faculty result = facultyService.addFaculty(facultyToAdd))

Antony8720 commented on src/test/java/ru/hogwarts/school/service/FacultyServiceImplTest.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    @Test
+    public void shouldCorrectlyFindFaculty() {
```
Для всех остальных тестов замечание выше также актуально

Boroda9992 committed:
Удалось поправить только тесты на добавление студента. Остальные тесты выбрасывают java.util.NoSuchElementException: No value present
В открытых источниках пока не нашел решения проблемы. С одной стороны, меня предостерегают от использования .get(), т.к. Mockito возвращает Optional с которым лучше .get() не использовать. Однако, как именно обойти этот момент в работе с репозиториями, я пока найти не смог.
Эксперименты ставил на тестах StudentServiceImplTest, набор тестов для факультета пока не трогал, но ошибку свою понимаю - не обращаюсь к репозиторию, а проверяю созданные руками объекты.
Вероятно, решение проще, чем я думаю, однако сам пока найти его так и не смог.
Так же буду благодарен, если подскажете ресурс, где можно найти примеры использования Mockito не только в самых примитивных случаях. А лучше - потренироваться на примерах.

Antony8720 commented on src/test/java/ru/hogwarts/school/service/StudentServiceImplTest.java:
```diff
@@ -0,0 +1,76 @@
+    private StudentServiceImpl studentService;
+
+    @BeforeEach
+    public void setup() {
```
Этот метод не нужен здесь, поскольку ты используешь бин через аннотацию @InjectMocks

Boroda9992 replied:
Ранее это был метод clear(), который убирал ранее созданные объекты перед каждым тестом. Вернуть его в прежнее состояние или же удалить полностью? Не приведет ли полное удаление к ситуации, когда каждый последующий тест будет просто плодить тестовые объекты?

Antony8720 replied:
Нужно удалить его полностью, бины spring по умолчанию создаются как синглтон, поэтому пересоздания объектов не будет

Antony8720 commented on src/test/java/ru/hogwarts/school/service/StudentServiceImplTest.java:
```diff
@@ -0,0 +1,76 @@
+//        //given
+//        Student studentToFind = new Student(1, "Harry", 12);
+//        //when
+//        Mockito.when(studentRepository.save(studentToFind)).thenReturn(studentToFind);
```
Здесь мокируется неправильный метод, нужно мокировать метод, который вызывается внутри метода сервиса, здесь это метод findById, поэтому нужно мокировать его: 
`Mockito.when(studentRepository.findById(1L)).thenReturn(Optional.of(studentToFind));`

На счет тренировки: нужно просто понять смысл работы и использования мокито, это на самом деле довольно простой инструмент, учебных примеров на своем коде должно хватить

Boroda9992 replied:
Логика была в том, что создав объект руками, я не записываю его в репозиторий, а просто в переменную. После этого у меня шла такая же строка, где я пытался замокать именно метод сервиса, который и пытался тестировать studentService.findStudent(long id). Отталкиваясь от этого, я считал, что любые манипуляции с моком репозитория - это подготовка для тестирования методов сервиса. Следовательно, если я пойду приведенным вами путем, то получится, что я тестирую не сервис, а встроенный метод репозитория.

Antony8720 replied:
Мокать необходимо не метод сервиса, а встроенные в него методы сторонних классов. В этом и смысл моков, заменять реальные зависимости класса на фейковые с заранее заданным поведением. То есть тот метод, который мокируется, это всегда зависимость класса, который мы хотим протестировать. В данном случае для методов StudentService, возможно, это не сильно понятно, поскольку методы сервиса просто вызывают методы репозитория и никакой дополнительной логики и обработки данных не привносят.

Boroda9992 committed:
Поправил тесты. Добавил проверки на null и верификацию вызовов методов.

Antony8720 commented on src/test/java/ru/hogwarts/school/service/FacultyServiceImplTest.java:
```diff
@@ -0,0 +1,70 @@
+
+    @BeforeEach
+    public void clear() {
+        facultyService = new FacultyServiceImpl(facultyRepository);
```
Этот метод не нужен

Boroda9992 replied:
Метод clear() удалил, однако, вы явно смотрите предыдущий комит. В последнем комите не осталось закомментированных строк. Да и в целом, номера строк не совпадают.

Antony8720 commented on src/test/java/ru/hogwarts/school/service/FacultyServiceImplTest.java:
```diff
@@ -0,0 +1,70 @@
+    public void shouldCorrectlyFindFaculty() {
+        //given
+        Faculty facultyToFind = new Faculty(1, "Testendor", "000FFF");
+        facultyService.addFaculty(facultyToFind);
```
В данном тесте необходимо использовать мокито, сейчас он не проверят нужной логики.

Boroda9992 replied:
Код из устаревшего комита. Сейчас тест выглядит так:
 @Test
    public void shouldCorrectlyFindFaculty() {
        //given
        Faculty facultyToFind = new Faculty(1, "Testendor", "000FFF");
        Mockito.when(facultyRepository.findById(1L)).thenReturn(Optional.of(facultyToFind));
        //when
        Faculty result = facultyService.findFaculty(facultyToFind.getId());
        //then
        Assertions.assertEquals(facultyToFind, result);
        Mockito.verify(facultyRepository, times(1)).findById(1L);
    }

Antony8720 commented on src/test/java/ru/hogwarts/school/service/FacultyServiceImplTest.java:
```diff
@@ -0,0 +1,70 @@
+    public void shouldCorrectlyEditFaculty() {
+        //given
+        Faculty facultyToEdit = new Faculty(1, "Testendor", "000FFF");
+        facultyService.addFaculty(facultyToEdit);
```
В данном тесте необходимо использовать мокито, сейчас он не проверят нужной логики.

Boroda9992 replied:
Код из старого комита.

Antony8720 commented on src/test/java/ru/hogwarts/school/service/FacultyServiceImplTest.java:
```diff
@@ -0,0 +1,70 @@
+        //given
+        Faculty facultyToDelete = new Faculty(1, "Testendor", "000FFF");
+        facultyService.addFaculty(facultyToDelete);
+        //when
```
В данном тесте необходимо использовать мокито, сейчас он не проверят нужной логики.

Boroda9992 replied:
Код из старого комита.

Antony8720 commented on .mvn/wrapper/maven-wrapper.properties:
```diff
@@ -0,0 +1,19 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
```
Этого файла не должно быть в репозитории, используй .gitignore

Boroda9992 replied:
На курсе нам не давали по .gitignore ничего, кроме исключения папок idea и файлов .iml... Учту, подчищу.

Antony8720 commented on mvnw:
```diff
@@ -0,0 +1,259 @@
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
```
Этого файла не должно быть в репозитории, используй .gitignore

Boroda9992 replied:
На курсе нам не давали по .gitignore ничего, кроме исключения папок idea и файлов .iml... Учту, подчищу.

Antony8720 commented on mvnw.cmd:
```diff
@@ -0,0 +1,149 @@
+@REM    http://www.apache.org/licenses/LICENSE-2.0
+@REM
+@REM Unless required by applicable law or agreed to in writing,
+@REM software distributed under the License is distributed on an
```
Этого файла не должно быть в репозитории, используй .gitignore

Boroda9992 replied:
На курсе нам не давали по .gitignore ничего, кроме исключения папок idea и файлов .iml... Учту, подчищу.

Boroda9992 committed:
- "clear()" удален из обоих тестовых веток.
- В ".gitignore" добавлены: ".mvn/wrapper/maven-wrapper.properties","mvnw","mvnw.cmd".

Boroda9992 committed:
- Добавил метод findByAgeBetween().
- Создал файл scripts.sql с 5-ю скриптами по заданию
- Обновил модели Студента и Факультета: добавил поля и аннотации @ManyToOne и @OneToMany соответственно.

Antony8720 commented on src/main/java/ru/hogwarts/school/controller/FacultyController.java:
```diff
@@ -43,5 +45,9 @@ public ResponseEntity<Void> deleteFaculty(@PathVariable long id) {
         facultyService.deleteFaculty(id);
         return ResponseEntity.ok().build();
     }
```
Не нашел этих эндпоинтов:

Добавить два эндпоинта в соответствующие контроллеры которые позволят:

1. Получить факультет студента
2. Получить студентов факультета

Antony8720 commented on src/main/java/ru/hogwarts/school/model/Student.java:
```diff
@@ -0,0 +1,81 @@
+    final String name;
+    final int age;
+    @ManyToOne
+    @JoinColumn(name = "faculty_id")
```
Нужно использовать FetchType = LAZY
https://javarush.com/quests/lectures/questhibernate.level14.lecture01

Antony8720 commented on src/main/java/ru/hogwarts/school/model/Faculty.java:
```diff
@@ -0,0 +1,88 @@
+    final String name;
+    final String colour;
+
+    @OneToMany(mappedBy = "faculty")
```
Нужно использовать FetchType = LAZY
https://javarush.com/quests/lectures/questhibernate.level14.lecture01

Antony8720 commented on src/main/java/ru/hogwarts/school/controller/FacultyController.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    @GetMapping("{id}")
+    public ResponseEntity<Faculty> getFacultyInfo(@PathVariable long id) {
```
Если поглубже изучить аннотацию @RestController, то она оборачивает (с помощью аннотации @ResponseBody) тело каждого метода в ResponseEntity. Таким образом нам нет смысла явно использовать этот класс в сигнатуре метода, и избавившись от него, наш код станет более удобочитаемым.

Boroda9992 committed:
- создал методы и эндпоинты для поиска всех студентов факультета и поиска факультета студента

Antony8720 commented on src/main/java/ru/hogwarts/school/controller/FacultyController.java:
```diff
@@ -0,0 +1,59 @@
+        return facultyService.addFaculty(faculty);
+    }
+
+    @PutMapping("{id}")
```
Мой прошлый комментарий не исправлен:
Если поглубже изучить аннотацию @RestController, то она оборачивает (с помощью аннотации @responsebody) тело каждого метода в ResponseEntity. Таким образом нам нет смысла явно использовать этот класс в сигнатуре метода, и избавившись от него, наш код станет более удобочитаемым.

Убери использование ResponseEntity из методов контроллера

Boroda9992 committed:
- убрал ResponseEntity из контроллеров

Boroda9992 committed:
- убрал ResponseEntity из контроллеров
